esi_leap/db/sqlalchemy/api.py

```diff
@@ -133,23 +133,139 @@ def offer_get_all(filters):
     start = filters.pop('start_time', None)
     end = filters.pop('end_time', None)
 
+    a_start = filters.pop('available_start_time', None)
+    a_end = filters.pop('available_end_time', None)
+
+    if a_start and a_end and a_end <= a_start:
+        raise exception.InvalidAvailabilityCommand(a_start=a_start,
+                                                   a_end=a_end)
+
     query = query.filter_by(**filters)
 
     if start and end:
         query = query.filter((start >= models.Offer.start_time) &
                              (end <= models.Offer.end_time))
 
+    for o in query:
+        try:
+            offer_verify_availability(o, start_time=a_start, end_time=a_end)
+        except exception.OfferNotAvailable:
+            query = query.filter(models.Offer.uuid != o.uuid)
+
     return query
 
 
+def offer_get_availabilities(offer_ref):
```

Comment:
Is there a reason not to do this entire function in `objects/offer.py`?

```
contracts = self.contracts ()    // if there isn't a method for this, there should be 
for contract in contracts:
    // construct list of availabilities
```

Reply:
This function gets used by other db functions so I didn't want the db calling object level code

Reply:
I think it's only used to determine default start/end times, correct? If we simplify that, we can move this out of the db layer.